ethmonitor/ethmonitor.go

```diff
@@ -287,33 +303,66 @@ func (m *Monitor) listenNewHead() <-chan uint64 {
 	var latestHeadBlock atomic.Uint64
 	nextBlock := make(chan uint64)
 
-	if m.provider.IsStreamingEnabled() {
-		// Streaming mode if available, where we listen for new heads
-		// and push the new block number to the nextBlock channel.
-		go func() {
-		reconnect:
+	go func() {
+		var streamingErrorCount int
+		var streamingErrorLastTime time.Time
+
+	reconnect:
+		// reset the latest head block
+		latestHeadBlock.Store(0)
+
+		// if we have too many streaming errors, we'll switch to polling
+		streamingErrorCount++
+		if time.Since(streamingErrorLastTime) > m.options.StreamingErrorResetInterval {
+			streamingErrorCount = 0
+		}
+
+		// listen for new heads either via streaming or polling
+		if m.provider.IsStreamingEnabled() && streamingErrorCount < m.options.StreamingErrNumToSwitchToPolling {
+			// Streaming mode if available, where we listen for new heads
+			// and push the new block number to the nextBlock channel.
+			m.log.Info("ethmonitor: starting stream head listener")
+
 			newHeads := make(chan *types.Header)
 			sub, err := m.provider.SubscribeNewHeads(m.ctx, newHeads)
 			if err != nil {
 				m.log.Warnf("ethmonitor: websocket connect failed: %v", err)
 				m.alert.Alert(context.Background(), "ethmonitor: websocket connect failed", err)
 				time.Sleep(1500 * time.Millisecond)
+
+				streamingErrorLastTime = time.Now()
 				goto reconnect
 			}
 
+			blockTimer := time.NewTimer(3 * m.options.ExpectedBlockInterval)
```

Comment:
Resource leak, agrh... defer will not help as it never goes out of function scope. I will get that fixed asap.

Reply:
https://github.com/0xsequence/ethkit/pull/119 -- hows this..?

Reply:
Actually, it is good. The timer will close itself after it fires and will be garbage collected. So there is no need to close it. It will also not cause process to hang. There will be at most a couple of them.

```golang
// After waits for the duration to elapse and then sends the current time
// on the returned channel.
// It is equivalent to NewTimer(d).C.
// The underlying Timer is not recovered by the garbage collector
// until the timer fires. If efficiency is a concern, use NewTimer
// instead and call Timer.Stop if the timer is no longer needed.
func After(d Duration) <-chan Time {
	return NewTimer(d).C
}
```

Reply:
“a little temporary timer leak” is fine imho

it’d probably be better to close it, though, to be accurate
(a leaked timeout might cause an unnecessary OS interrupt)